src/Database/LSMTree/Internal/MergeSchedule.hs

```diff
@@ -908,16 +914,30 @@ supplyMergeCredits _ SingleRun{} = pure ()
 supplyMergeCredits (ScaledCredits c) (MergingRun _ _ var) = do
     mergeIsDone <- withMVar var $ \case
       CompletedMerge{} -> pure False
-      (OngoingMerge _rs pvar m) -> do
-        -- n >= c
-        (n, stepResult) <- Merge.steps m c
-        x <- readPrimVar pvar
-        writePrimVar pvar $! n + x
+      (OngoingMerge _rs totalStepsVar totalCreditsVar m) -> do
+        totalSteps <- readPrimVar totalStepsVar
+        totalCredits <- readPrimVar totalCreditsVar
+
+        -- If we previously performed too many merge steps, then we perform
+        -- fewer now.
+        let stepsToDo = totalCredits - totalSteps + c
```

Comment:
Yes ok I see the point. I'm trying to convince myself that this is the right way round and that `stepsToDo` is guaranteed positive.

The totalSteps is >= the totalCredits right, because Merge.steps guarantees to do at least the number of steps requested, but perhaps more.

So, isn't `totalCredits - totalSteps` backwards? That's typically going to be negative.

Don't we want something like `max 0 (totalCredits + c - totalSteps)`?

BTW, we might as well assert that `totalSteps >= the totalCredits` and 'stepsDone >= stepsToDo'.

Reply:
Yes, we should probably do a least `0` steps

Reply:
Or skip the rest of this body entirely when `stepsToDo <= 0`